getline() not available on OSX pre-10.7

# Context

* Version of iperf3:
4f3a7a5403b61f49ab277d0756e87c651a45bdfe

* Hardware:
Mac

* Operating system (and distribution, if any):
Snow Leopard (10.6) and prior

* Other relevant information (for example, non-default compilers,
  libraries, cross-compiling, etc.):
built with MacPorts (gcc 4.2)

# Bug Report

* Expected Behavior
successful build

* Actual Behavior
```
Undefined symbols:
  "_getline", referenced from:
      _iperf_getpass in iperf3_profile-iperf_auth.o
ld: symbol(s) not found
```

* Steps to Reproduce
N/A

* Possible Solution
Replace the missing getline() function in [src/iperf_auth.c](https://github.com/esnet/iperf/blob/master/src/iperf_auth.c) for OSX <= 10.6. This was [proposed to libgss](http://lists.gnu.org/archive/html/help-gss/2017-04/msg00001.html) and that patch can be used here as well.

@bmah888, I'm afraid I'm out of my league there. I submitted a PR to test building this patch and it still failed. Below is the relevant portion from the output (10.6 x86_64) and it looks like `iperf_auth.c` (first line) is built before `iperf_util.c` (last line) got to define the `getline()` function. Here is [the full log](https://build.macports.org/builders/ports-10.6_x86_64_legacy-builder/builds/34864/steps/install-port/logs/stdio) if you need it. Is that correct?

```
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_auth.lo -MD -MP -MF .deps/iperf_auth.Tpo -c iperf_auth.c  -fno-common -DPIC -o .libs/iperf_auth.o
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_client_api.lo -MD -MP -MF .deps/iperf_client_api.Tpo -c iperf_client_api.c  -fno-common -DPIC -o .libs/iperf_client_api.o
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_locale.lo -MD -MP -MF .deps/iperf_locale.Tpo -c iperf_locale.c -o iperf_locale.o >/dev/null 2>&1
mv -f .deps/iperf_locale.Tpo .deps/iperf_locale.Plo
/bin/sh ../libtool  --tag=CC   --mode=compile /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I.   -I/usr/include -I/opt/local/include  -pipe -Os -arch x86_64 -Wall -MT iperf_udp.lo -MD -MP -MF .deps/iperf_udp.Tpo -c -o iperf_udp.lo iperf_udp.c
iperf_api.c: In function ‘iperf_print_intermediate’:
iperf_api.c:2482: warning: format ‘%lu’ expects type ‘long unsigned int’, but argument 3 has type ‘iperf_size_t’
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_error.lo -MD -MP -MF .deps/iperf_error.Tpo -c iperf_error.c -o iperf_error.o >/dev/null 2>&1
iperf_auth.c: In function ‘Base64Encode’:
iperf_auth.c:112: warning: value computed is not used
iperf_auth.c:114: warning: value computed is not used
iperf_auth.c: In function ‘iperf_getpass’:
iperf_auth.c:294: warning: implicit declaration of function ‘getline’
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_tcp.lo -MD -MP -MF .deps/iperf_tcp.Tpo -c iperf_tcp.c -o iperf_tcp.o >/dev/null 2>&1
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_client_api.lo -MD -MP -MF .deps/iperf_client_api.Tpo -c iperf_client_api.c -o iperf_client_api.o >/dev/null 2>&1
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_udp.lo -MD -MP -MF .deps/iperf_udp.Tpo -c iperf_udp.c  -fno-common -DPIC -o .libs/iperf_udp.o
mv -f .deps/iperf_error.Tpo .deps/iperf_error.Plo
/bin/sh ../libtool  --tag=CC   --mode=compile /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I.   -I/usr/include -I/opt/local/include  -pipe -Os -arch x86_64 -Wall -MT iperf_sctp.lo -MD -MP -MF .deps/iperf_sctp.Tpo -c -o iperf_sctp.lo iperf_sctp.c
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_server_api.lo -MD -MP -MF .deps/iperf_server_api.Tpo -c iperf_server_api.c -o iperf_server_api.o >/dev/null 2>&1
libtool: compile:  /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I. -I/usr/include -I/opt/local/include -pipe -Os -arch x86_64 -Wall -MT iperf_auth.lo -MD -MP -MF .deps/iperf_auth.Tpo -c iperf_auth.c -o iperf_auth.o >/dev/null 2>&1
mv -f .deps/iperf_tcp.Tpo .deps/iperf_tcp.Plo
/bin/sh ../libtool  --tag=CC   --mode=compile /usr/bin/gcc-4.2 -DHAVE_CONFIG_H -I.   -I/usr/include -I/opt/local/include  -pipe -Os -arch x86_64 -Wall -MT iperf_util.lo -MD -MP -MF .deps/iperf_util.Tpo -c -o iperf_util.lo iperf_util.c
```

--- a/configure.ac
+++ b/configure.ac
@@ -160,6 +160,10 @@ AC_CHECK_FUNCS([cpuset_setaffinity sched_setaffinity],
 # it needs and what arguments it expects.
 AC_CHECK_FUNCS([sendfile])
 
+# Check for getline support, used as a part of authenticated
+# connections.
+AC_CHECK_FUNCS([getline])
+
 # Check for packet pacing socket option (Linux only for now).
 AC_CACHE_CHECK([SO_MAX_PACING_RATE socket option],
 [iperf3_cv_header_so_max_pacing_rate],

--- a/src/iperf_util.c
+++ b/src/iperf_util.c
@@ -432,3 +432,86 @@ iperf_dump_fdset(FILE *fp, char *str, int nfds, fd_set *fds)
     }
     fprintf(fp, "]\n");
 }
+
+/* Compatibility version of getline(3) for systems that don't have it.. */
+#ifndef HAVE_GETLINE
+/* The following code adopted from NetBSD's getline.c, which is: */
+
+/*-
+ * Copyright (c) 2011 The NetBSD Foundation, Inc.
+ * All rights reserved.
+ *
+ * This code is derived from software contributed to The NetBSD Foundation
+ * by Christos Zoulas.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
+ * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+ssize_t
+getdelim(char **buf, size_t *bufsiz, int delimiter, FILE *fp)
+{
+	char *ptr, *eptr;
+
+
+	if (*buf == NULL || *bufsiz == 0) {
+		*bufsiz = BUFSIZ;
+		if ((*buf = malloc(*bufsiz)) == NULL)
+			return -1;
+	}
+
+	for (ptr = *buf, eptr = *buf + *bufsiz;;) {
+		int c = fgetc(fp);
+		if (c == -1) {
+			if (feof(fp)) {
+				ssize_t diff = (ssize_t)(ptr - *buf);
+				if (diff != 0) {
+					*ptr = '\0';
+					return diff;
+				}
+			}
+			return -1;
+		}
+		*ptr++ = c;
+		if (c == delimiter) {
+			*ptr = '\0';
+			return ptr - *buf;
+		}
+		if (ptr + 2 >= eptr) {
+			char *nbuf;
+			size_t nbufsiz = *bufsiz * 2;
+			ssize_t d = ptr - *buf;
+			if ((nbuf = realloc(*buf, nbufsiz)) == NULL)
+				return -1;
+			*buf = nbuf;
+			*bufsiz = nbufsiz;
+			eptr = nbuf + nbufsiz;
+			ptr = nbuf + d;
+		}
+	}
+}
+
+ssize_t
+getline(char **buf, size_t *bufsiz, FILE *fp)
+{
+	return getdelim(buf, bufsiz, '\n', fp);
+}
+
+#endif

